File: php/token_generator.php
<?php

    function generate_token($resource, $generator, $secret_key) {
        /* accepts arguments:
           $resource = a resource on a web server e.g. /path/to/resource?option=33
           $generator = an integer generator value
           $secret_key = the secret key for the HMAC construction.

           Returns the concaternation of the generator and the first 20 characters 
           of the HMAC of $resource using key $secret_key and the SHA1 algorithm.
         */
        $hmac_str = hash_hmac('sha1', $resource, $secret_key);
        return $generator . substr($hmac_str, 0, 20);
    }

    function construct_url_with_token($resource, $generator, $secret) {
        /* construct_url_with token takes arguments:
           $resource = a resource on a web server e.g. /path/to/resource?option=33
           $generator = an integer generator value
           $secret_key = the secret key for the HMAC construction.

           This will append &encoded=token to your resource url
         */
        $token = generate_token($resource, $generator, $secret);
        $new_resource = $resource . "&encoded=" . $token;
        return $new_resource;
    }


    /*
     * This block ensures the code running here only runs 
     * if you execute this file directly using PHP.
     */
    if (!count(debug_backtrace())) {
        /* you can use this snippet in your own code if you
           include 'path/to/token_generator.php';
         */
        $g = 0;
        $key = "There's no place like home!";
        $uri = "/path/to/resource?sessionid=12345&misc=abcde&stime=1&etime=9000000000";
        $expected_outcome = "/path/to/resource?sessionid=12345&misc=abcde&stime=1&etime=9000000000&encoded=0760514fbdf916646f69b";
        $uri_plus_token = construct_url_with_token($uri, $g, $key);

        /* you don't need this - this is to verify the code produces the 
           expected result */
        print("Result is " . $uri_plus_token . "\n");
        assert($uri_plus_token == $expected_outcome);
    }

?>

File: python/token_generator.py
#!/usr/bin/env python
"""
Swiftserve 2 Token Generation code.

You can execute this file using python token_generator.py
and you will see an example printout of how it works.
"""

import sys
import hashlib
import hmac
import re


def generate_token(resource, generator, secret_key):
    """
    Takes arguments:
    resource = a string representing a http resource, e.g. /resource?option = 1
    generator = a (random) integer value.
    secret_key = a string containing the secret key

    Calculates the HMAC_SHA1 of the resource, using the secret_key as the key
    and prepends the generator value.
    """
    assert(type(generator) == int)
    hmac_builder = hmac.new(secret_key, resource, hashlib.sha1)
    #print hmac_builder.hexdigest()
    token_string = "%d%s" % (generator, hmac_builder.hexdigest()[:20])
    return token_string


def construct_url_with_token(resource, generator, secret_key):
    """
    Takes arguments:
    resource = a string representing a http resource, e.g. /resource?option=1
    generator = a (random) integer value.
    secret_key = a string containing the secret key

    Appends &encoded=token to the resource given.
    """
    assert(type(generator) == int)
    token = generate_token(resource, generator, secret_key)
    new_resource = "%s&encoded=%s" % (resource, token)
    return new_resource


"""
This construction allows you to demo this code using a preset example
you can use this code directly with:

    from token_generator import construct_url_with_token

Please *note*: SS2 authentication requires the resource to contain
mandatory `stime' and `etime' arguments in the query string
part of the resource specifier.

"""

def demo():
    g = 0
    key = "odiAS3r98aoinfFD383Orpt346SDFDG3r9ef92gjdgoi"

    uri = "/clip.mp4?stime=1322079229&etime=1322089229"
    expected_outcome = "/clip.mp4?stime=1322079229&etime=1322089229&encoded=0c5e4cc33315fa2ab5465"

    #uri = "/clip.mp4/manifest.f4m?stime=1322079229&etime=1322089229"
    #expected_outcode = "/clip.mp4/manifest.f4m?stime=1322079229&etime=1322089229&encoded=053bfdb975e4c18024647"

    uri_plus_token = construct_url_with_token(uri, g, key)

    print "This is a demonstraton of token generation."
    print "Original URL: ", uri
    print "Tokenised URL: ", uri_plus_token
    print "Expected outcome: ", expected_outcome
    assert(uri_plus_token == expected_outcome)

if __name__ == '__main__':
    g = 0

    try:
        key, uri = sys.argv[1:3]
    except ValueError:
        demo()
        sys.exit(0)

    print construct_url_with_token(uri, g, key)
